rfc-115-enabling-http2-on-govuk.md

```diff
@@ -0,0 +1,123 @@
+
+Once completed our connection graph will look similar to this:
+
+![Example of an optimised connection graph](rfc-115/clean-connections.png)
+
+The bulk of the page assets will be downloaded on the initial connection to the origin (www.gov.uk), with a secondary `anonymous` connection only opened for the fonts. That's twelve connections under our current HTTP/1.1 implementation, down to just two for HTTP/2 in many cases. So in order to fix our HTTP/1.1 and HTTP/2 performance issues we should serve static assets (CSS, JavaScript, fonts, images) all from the origin (www.gov.uk).
+
+### Summary
+In order to enable HTTP/2 on the GOV.UK domain (and for it to perform well) we need to consider problems 2 and 3 listed above:
+
+- SRI requires the `crossorigin` attribute to be set for it to work.
+- The use of `crossorigin='anonymous'` breaks HTTP/1.1 connection reuse and HTTP/2 connection coalescing.
+- Switching to use `crossorigin='use-credentials'` isn't possible because we serve our assets with the `Access-Control-Allow-Origin: *` header so that the assets can be used on other domains (e.g. Heroku).
+- Even if we serve the static assets from the origin, we will still need the `crossorigin` attribute because we will still be loading them with SRI enabled.
+
+Therefore my recommendation is to remove the `integrity` and `crossorigin` attributes from our static resources (CSS, JS) and work towards serving these assets from the origin domain so we are no longer reliant on an individual browsers implementation of HTTP/2 connection coalescing. NOTE: it is possible to do these streams of work separately. 
+
+SRI Note: In serving the static assets from the origin domain, SRI no longer serves any purpose, as it's a security measure intended to be used on third party resources. As all of our assets are first party, it really isn't required and serves only to complicate our setup.
+
+Serving static assets from the origin also clears a path for future web performance enhancements like [HTTP/3, QUIC (Quick UDP Internet Connection)](https://www.fastly.com/blog/why-fastly-loves-quic-http3) and 0-RTT. [HTTP/3](https://caniuse.com/#feat=http3) has no support at the moment as it currently sits behind browser flags. But at some point in the future it will be supported and the asset domain will need to be removed for static assets anyway to keep up with the latest protocol developments.
+
+NOTE: As mentioned by @david-ncsc in [his comment](https://github.com/alphagov/govuk-rfcs/pull/115#issuecomment-567907397), there's probably a security benefit to keeping the uploaded files on a separate origin as defence against uploaded malicious files. We should therefore only serve the static assets (CSS, JavaScript, fonts) from the origin. All other assets can stay the same.
+
+## Proposal
+### MUST
+* Continue to serve all our assets with `Access-Control-Allow-Origin: *` header. - Easy (no work required)
```

Comment:
Is this still a must if we drop the SRI tags? I'd have thought in that scenario it's only needed for fonts

Reply:
This header is still required to allow our Heroku applications to be allowed to fetch CSS assets from the assets domain. Without it the CORS will kick in and block the fetch.

Reply:
We're proposing that they're switched to not use SRI right? So we'd have the equivalent of `<link rel="stylesheet" href="https://www.gov.uk/assets/static/application.css">`  and `<script src="https://www.gov.uk/assets/static/application.js"></script>` right?

In that case I don't believe there are any CORS restrictions that apply (unless I've fallen out of touch with the web again - which is a real possibility).

Reply:
Yes that's correct, there are no restrictions on the GOV.UK side, as all assets will be coming from the same domain. Including the header is stopping our Heroku apps from breaking.

Reading into CORS:
> Cross-Origin Resource Sharing (CORS) is a mechanism that uses additional HTTP headers to tell browsers to give a web application running at one origin, access to selected resources from a different origin.

So my interpretation of that is Heroku will be making a cross-origin request to assets sitting on the assets domain. And for a browser to be allowed to do that we either include:

- `Access-Control-Allow-Origin: *` 
- `Access-Control-Allow-Origin: https://[example-app].heroku.com`

We haven't come across this issue with Heroku before because we've always included the `Access-Control-Allow-Origin: *` header when serving assets from the assets domain.  
If these aren't included then the browser will stop the fetch, and no JS (and potentially CSS) will be seen on our Heroku apps.

Reply:
Due to the lack of SRI those links and scripts would continue to work for apps on any hostname too. The need for the `Access-Control-Allow-Origin: *` header only kicks in when we add the crossorigin attribute.

Reply:
CORS breaks my brain. Sure that makes sense, although reading around I seem to be finding mixed ideas on what should / does happen. 

I'll rewrite the RFC so that we are only serving fonts with `Access-Control-Allow-Origin: *`. It sounds we only need the `crossorigin` on scripts if we require information about the error information of the script being loaded (or SRI as you mention).